### PR TITLE
Generate `Expr(:incomplete)`

### LINKED ISSUE #42: When using in REPL, `JuliaSyntax.jl` does not seem to work well with multiline parsing?
When I want to type

```julia
map(1:10) do x
    2x
end
```

after I type `map(1:10) do x` and hit return, I got

```julia
julia> map(1:10) do
ERROR: ParseError:
Error: premature end of input
@ REPL[11]:2:1
map(1:10) do
Error: premature end of input
@ REPL[11]:2:1
map(1:10) do
Error: Expected `end` but got unexpected tokens
@ REPL[11]:2:1
map(1:10) do
```

Or if I want to type

```julia
 w = u +
     v +
     1
```

I got

```julia
julia> w = u +
ERROR: ParseError:
Error: premature end of input
@ REPL[14]:2:1
w = u +
```

But I can do them with copy-paste:

image

Can we make the REPL-mode tolerate multiline code a little bit more?

### PR DESCRIPTION
**Pull request by c42f**
This allows REPL completion to work correctly. It works by pattern matching the parse tree, rather than hard coding incomplete expression detection into the parser itself.

This took a while because I got distracted with #88. Though doing some of those things first helped make this quite a lot cleaner. There's still more changes from #88 which would help make this nicer but for now it works.

The tests here are somewhat derived from `Base`, but I had to review them all (ie the tests pasted in https://github.com/JuliaLang/JuliaSyntax.jl/pull/71#issuecomment-1225364445) because they turn out to not really make sense.  For example, `"begin;"` and `"begin"` are both the prefix of a block construct, one of them shouldn't come out as `:other`.

Alternative to #71, CC @aviatesk :-)

Fix #42, would close #71

### CODE REVIEW

File: test/hooks.jl [hunk 1]
```diff
@@ -31,8 +31,68 @@
...
+                "mutable struct"    => :other
+                "mutable struct X"  => :block
+                "struct"       => :other
+                "struct X"     => :block
+                "if"           => :other
+                "if x"         => :block
+                "while"        => :other
+                "while x"      => :block
+                "try"          => :block
+                "try x catch"  => :block  # ambiguous - could be :block or :other
```
Comment (pfitzseb) [inline test/hooks.jl:79]: IMHO `:block` is correct here, since `try x catch end` is valid.

Comment (c42f) [reply test/hooks.jl:79]: I thought `:other` also fits because `try x catch exc body end` parses. So could either be a block or a symbol for the exception name.  So this seemed ambiguous to me?

Comment (c42f) [reply test/hooks.jl:79]: (BTW, I don't think the REPL does much with the distinction here. Though the REPL completion code is a bit hard to follow.)

Comment (pfitzseb) [reply test/hooks.jl:79]: `:other` doesn't make sense to me since the presence of `exc` is irrelevant for the validity of the code, but I also don't really know how this value actually ends up being used.

Comment (c42f) [reply test/hooks.jl]: Seems fair, I'll just change the note next to this to explain the possibility of both syntaxes.  AFAIK this doesn't change things much either way with the REPL (and in any case as noted in the description, I think `Base.incomplete_tag` itself is quite inconsistent in any case).
